src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -143,101 +371,179 @@ check if we have not found a block (when block == null
         }
-        //get length or width either is larger.
-        final int length = currentStructure.getLength();
-        final int width = currentStructure.getWidth();
-        final int distance = Math.max(width, length) + MIN_ADDITIONAL_RANGE_TO_BUILD + offset;
-        @NotNull final EnumFacing[] directions = {EnumFacing.EAST, EnumFacing.WEST, EnumFacing.NORTH, EnumFacing.SOUTH};
+    }
 
-        //then get a solid place with two air spaces above it in any direction.
-        for (final EnumFacing direction : directions)
+    private void requestMaterials()
+    {
+        //We need to deal with materials
+        if (!Configurations.builderInfiniteResources && job instanceof JobBuilder)
         {
-            @NotNull final BlockPos positionInDirection = getPositionInDirection(direction, distance);
-            if (EntityUtils.checkForFreeSpace(world, positionInDirection))
+            while (((JobBuilder) job).getStructure().findNextBlock())
             {
-                return positionInDirection;
+                @Nullable final Block block = ((JobBuilder) job).getStructure().getBlock();
+                @NotNull final IBlockState blockState = ((JobBuilder) job).getStructure().getBlockState();
+
+                if (((JobBuilder) job).getStructure().doesStructureBlockEqualWorldBlock() ||
+                        (blockState instanceof BlockBed && blockState.getValue(BlockBed.PART).equals(BlockBed.EnumPartType.FOOT))
+                        || (blockState instanceof BlockDoor && blockState.getValue(BlockDoor.HALF).equals(BlockDoor.EnumDoorHalf.LOWER)))
+                {
+                    continue;
+                }
+
+                final Block worldBlock = BlockPosUtil.getBlock(world, ((JobBuilder) job).getStructure().getBlockPosition());
+
+                if (block != null
+                        && block != Blocks.AIR
+                        && worldBlock != Blocks.BEDROCK
+                        && !(worldBlock instanceof AbstractBlockHut)
+                        && !isBlockFree(block, 0))
+                {
+                    final AbstractBuilding building = getOwnBuilding();
+                    if (building instanceof BuildingBuilder)
+                    {
+                        ((BuildingBuilder) building).addNeededResource(block, 1);
+                    }
+                }
             }
+            ((JobBuilder) job).getWorkOrder().setRequested(true);
         }
-
-        //if necessary we can could implement calling getWorkingPosition recursively and add some "offset" to the sides.
-        return getWorkingPosition(offset + 1);
     }
 
+
     /**
      * Works on clearing the area of unneeded blocks.
      *
      * @return the next step once done.
      */
-    private boolean clearStep(@NotNull final Structure.SchematicBlock currentBlock)
+    private boolean clearStep(@NotNull final Structure.StructureBlock currentBlock)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "clearStep" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)